File: wfm-develop mobile 07.05.2026/wfm/mobile/ios/WFMApp/Core/Networking/CacheManager.swift
import Foundation
import os.log

/// Управляет кэшированием HTTP ответов (in-memory + disk)
actor CacheManager {
    private let logger = Logger(subsystem: "com.wfm", category: "CacheManager")

    // TTL кэша (7 дней)
    private let defaultTTL: TimeInterval = 60 * 60 * 24 * 7

    // In-memory cache
    private let memoryCache = NSCache<NSString, CacheEntry>()

    // Disk cache directory
    private let diskCacheURL: URL

    // Максимальный размер in-memory cache (50 MB)
    private let memoryCacheLimit = 50 * 1024 * 1024

    // Максимальный размер disk cache (100 MB)
    private let diskCacheLimit = 100 * 1024 * 1024

    init() {
        // Настройка in-memory cache
        memoryCache.totalCostLimit = memoryCacheLimit

        // Настройка disk cache
        let cacheDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first!
        diskCacheURL = cacheDirectory.appendingPathComponent("com.wfm.http-cache", isDirectory: true)

        // Создаём директорию если не существует
        try? FileManager.default.createDirectory(at: diskCacheURL, withIntermediateDirectories: true)

        logger.info("📦 CacheManager initialized (TTL: \(self.defaultTTL)s, Memory: \(self.memoryCacheLimit / 1024 / 1024)MB, Disk: \(self.diskCacheLimit / 1024 / 1024)MB)")
    }

    /// Получить данные из кэша
    func get<T: Decodable>(key: String) -> T? {
        // 1. Проверяем in-memory cache
        if let entry = memoryCache.object(forKey: key as NSString) {
            // Проверяем не истёк ли TTL
            if Date().timeIntervalSince(entry.timestamp) < defaultTTL {
                logger.debug("✅ Memory cache HIT: \(key)")
                do {
                    return try JSONDecoder().decode(T.self, from: entry.data)
                } catch is DecodingError {
                    // Схема данных изменилась - удаляем из кэша
                    logger.warning("⚠️ Schema changed, clearing memory cache for: \(key)")
                    memoryCache.removeObject(forKey: key as NSString)
                    // Также удаляем с диска
                    let fileURL = diskCacheURL.appendingPathComponent(key.sha256())
                    try? FileManager.default.removeItem(at: fileURL)
                    return nil
                } catch {
                    // Другие ошибки - не удаляем кэш
                    logger.error("❌ Failed to decode from memory cache: \(error.localizedDescription)")
                    return nil
                }
            } else {
                logger.debug("⏰ Memory cache EXPIRED: \(key)")
                memoryCache.removeObject(forKey: key as NSString)
            }
        }

        // 2. Проверяем disk cache
        let fileURL = diskCacheURL.appendingPathComponent(key.sha256())

        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            logger.debug("❌ Cache MISS: \(key)")
            return nil
        }

        do {
            let data = try Data(contentsOf: fileURL)
            let entry = try JSONDecoder().decode(CacheEntry.self, from: data)

            // Проверяем не истёк ли TTL
            if Date().timeIntervalSince(entry.timestamp) < defaultTTL {
                logger.debug("✅ Disk cache HIT: \(key)")

                // Загружаем в memory cache
                memoryCache.setObject(entry, forKey: key as NSString, cost: entry.data.count)

                do {
                    return try JSONDecoder().decode(T.self, from: entry.data)
                } catch is DecodingError {
                    // Схема данных изменилась - удаляем из кэша
                    logger.warning("⚠️ Schema changed, clearing cache for: \(key)")
                    memoryCache.removeObject(forKey: key as NSString)
                    try? FileManager.default.removeItem(at: fileURL)
                    return nil
                }
            } else {
                logger.debug("⏰ Disk cache EXPIRED: \(key)")
                try? FileManager.default.removeItem(at: fileURL)
            }
        } catch is DecodingError {
            // Схема CacheEntry изменилась или данные повреждены - удаляем
            logger.warning("⚠️ Cache entry corrupted, clearing: \(key)")
            memoryCache.removeObject(forKey: key as NSString)
            try? FileManager.default.removeItem(at: fileURL)
        } catch {
            // Другие ошибки (например IO при чтении файла) - удаляем файл но это не связано со схемой
            logger.error("❌ Failed to read from disk cache: \(error.localizedDescription)")
            try? FileManager.default.removeItem(at: fileURL)
        }

        logger.debug("❌ Cache MISS: \(key)")
        return nil
    }

    /// Сохранить данные в кэш
    func set<T: Encodable>(key: String, value: T) {
        do {
            let data = try JSONEncoder().encode(value)
            let entry = CacheEntry(data: data, timestamp: Date())

            // 1. Сохраняем в memory cache
            memoryCache.setObject(entry, forKey: key as NSString, cost: data.count)

            // 2. Сохраняем на disk
            let fileURL = diskCacheURL.appendingPathComponent(key.sha256())
            let entryData = try JSONEncoder().encode(entry)
            try entryData.write(to: fileURL)

            logger.debug("💾 Cached: \(key) (\(data.count) bytes)")

            // 3. Проверяем размер disk cache и чистим если превышен лимит (в фоне)
            _Concurrency.Task {
                await cleanupDiskCacheIfNeeded()
            }
        } catch {
            logger.error("❌ Failed to cache: \(key) - \(error.localizedDescription)")
        }
    }

    /// Очистить весь кэш
    func clearAll() {
        // Очищаем memory cache
        memoryCache.removeAllObjects()

        // Очищаем disk cache
        try? FileManager.default.removeItem(at: diskCacheURL)
        try? FileManager.default.createDirectory(at: diskCacheURL, withIntermediateDirectories: true)

        logger.info("🗑️ Cache cleared")
    }

    /// Удалить конкретную запись из кэша
    func remove(key: String) {
        // Удаляем из memory cache
        memoryCache.removeObject(forKey: key as NSString)

        // Удаляем с disk
        let fileURL = diskCacheURL.appendingPathComponent(key.sha256())
        try? FileManager.default.removeItem(at: fileURL)

        logger.debug("🗑️ Removed from cache: \(key)")
    }

    /// Проверить и очистить disk cache если превышен лимит
    private func cleanupDiskCacheIfNeeded() async {
        do {
            let files = try FileManager.default.contentsOfDirectory(
                at: self.diskCacheURL,
                includingPropertiesForKeys: [.fileSizeKey, .contentModificationDateKey],
                options: .skipsHiddenFiles
            )

            // Подсчитываем общий размер
            var totalSize = 0
            var fileInfos: [(url: URL, size: Int, date: Date)] = []

            for fileURL in files {
                let attributes = try FileManager.default.attributesOfItem(atPath: fileURL.path)
                let size = attributes[.size] as? Int ?? 0
                let date = attributes[.modificationDate] as? Date ?? Date.distantPast

                totalSize += size
                fileInfos.append((url: fileURL, size: size, date: date))
            }

            // Если превышен лимит - удаляем самые старые файлы
            if totalSize > self.diskCacheLimit {
                self.logger.warning("⚠️ Disk cache exceeded limit (\(totalSize / 1024 / 1024)MB / \(self.diskCacheLimit / 1024 / 1024)MB)")

                // Сортируем по дате (самые старые первые)
                fileInfos.sort { $0.date < $1.date }

                var sizeToRemove = totalSize - self.diskCacheLimit
                var removedCount = 0

                for fileInfo in fileInfos {
                    if sizeToRemove <= 0 { break }

                    try? FileManager.default.removeItem(at: fileInfo.url)
                    sizeToRemove -= fileInfo.size
                    removedCount += 1
                }

                self.logger.info("🗑️ Removed \(removedCount) old cache entries")
            }
        } catch {
            self.logger.error("❌ Failed to cleanup disk cache: \(error.localizedDescription)")
        }
    }

    /// Генерирует ключ кэша из URL и query params
    static func cacheKey(path: String, queryItems: [URLQueryItem]?) -> String {
        var key = path

        if let queryItems = queryItems, !queryItems.isEmpty {
            let sortedQuery = queryItems
                .sorted { $0.name < $1.name }
                .map { "\($0.name)=\($0.value ?? "")" }
                .joined(separator: "&")
            key += "?" + sortedQuery
        }

        return key
    }
}

/// Запись кэша (данные + timestamp)
private class CacheEntry: Codable {
    let data: Data
    let timestamp: Date

    init(data: Data, timestamp: Date) {
        self.data = data
        self.timestamp = timestamp
    }
}

// MARK: - String Extension для SHA256
extension String {
    func sha256() -> String {
        // Простая хэш-функция для генерации имени файла
        // В production лучше использовать CryptoKit
        return String(self.hashValue)
    }
}

File: wfm-develop mobile 07.05.2026/wfm/mobile/android/app/src/main/java/com/beyondviolet/wfm/core/network/CacheManager.kt
package com.beyondviolet.wfm.core.network

import android.content.Context
import android.util.Log
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.util.concurrent.ConcurrentHashMap

/**
 * Управляет кэшированием HTTP ответов (in-memory + disk)
 */
class CacheManager(context: Context) {
    companion object {
        @PublishedApi
        internal const val TAG = "CacheManager"

        // TTL кэша (7 дней в миллисекундах)
        @PublishedApi
        internal const val DEFAULT_TTL = 7L * 24 * 60 * 60 * 1000

        // Максимальный размер disk cache (50 MB)
        @PublishedApi
        internal const val DISK_CACHE_LIMIT = 50L * 1024 * 1024

        /**
         * Генерирует ключ кэша из path и query params
         */
        fun cacheKey(path: String, queryParams: Map<String, String> = emptyMap()): String {
            if (queryParams.isEmpty()) {
                return path
            }

            val sortedQuery = queryParams.entries
                .sortedBy { it.key }
                .joinToString("&") { "${it.key}=${it.value}" }

            return "$path?$sortedQuery"
        }
    }

    // In-memory cache
    @PublishedApi
    internal val memoryCache = ConcurrentHashMap<String, CacheEntry>()

    // Disk cache directory
    @PublishedApi
    internal val diskCacheDir: File = context.cacheDir.resolve("http-cache").apply {
        if (!exists()) {
            mkdirs()
            Log.i(TAG, "📦 Created cache directory: $absolutePath")
        }
    }

    init {
        Log.i(TAG, "📦 CacheManager initialized (TTL: ${DEFAULT_TTL / 1000}s, Disk: ${DISK_CACHE_LIMIT / 1024 / 1024}MB)")
    }

    /**
     * Получить данные из кэша
     */
    inline fun <reified T> get(key: String): T? {
        // 1. Проверяем in-memory cache
        memoryCache[key]?.let { entry ->
            if (System.currentTimeMillis() - entry.timestamp < DEFAULT_TTL) {
                Log.d(TAG, "✅ Memory cache HIT: $key")
                return try {
                    Json.decodeFromString<T>(entry.data)
                } catch (e: kotlinx.serialization.SerializationException) {
                    // Схема данных изменилась - удаляем из кэша
                    Log.w(TAG, "⚠️ Schema changed, clearing memory cache for: $key")
                    memoryCache.remove(key)
                    // Также удаляем с диска
                    diskCacheDir.resolve(key.hashCode().toString()).delete()
                    null
                } catch (e: Exception) {
                    // Другие ошибки (например IOException) - не удаляем кэш
                    Log.e(TAG, "❌ Failed to decode from memory cache: ${e.message}")
                    null
                }
            } else {
                Log.d(TAG, "⏰ Memory cache EXPIRED: $key")
                memoryCache.remove(key)
            }
        }

        // 2. Проверяем disk cache
        val cacheFile = diskCacheDir.resolve(key.hashCode().toString())
        if (cacheFile.exists()) {
            try {
                val entryJson = cacheFile.readText()
                val entry = Json.decodeFromString<CacheEntry>(entryJson)

                if (System.currentTimeMillis() - entry.timestamp < DEFAULT_TTL) {
                    Log.d(TAG, "✅ Disk cache HIT: $key")

                    // Загружаем в memory cache
                    memoryCache[key] = entry

                    return try {
                        Json.decodeFromString<T>(entry.data)
                    } catch (e: kotlinx.serialization.SerializationException) {
                        // Схема данных изменилась - удаляем из кэша
                        Log.w(TAG, "⚠️ Schema changed, clearing cache for: $key")
                        memoryCache.remove(key)
                        cacheFile.delete()
                        null
                    }
                } else {
                    Log.d(TAG, "⏰ Disk cache EXPIRED: $key")
                    cacheFile.delete()
                }
            } catch (e: kotlinx.serialization.SerializationException) {
                // Схема CacheEntry изменилась или данные повреждены - удаляем
                Log.w(TAG, "⚠️ Cache entry corrupted, clearing: $key")
                memoryCache.remove(key)
                cacheFile.delete()
            } catch (e: Exception) {
                // Другие ошибки (например IOException при чтении файла) - удаляем файл но это не связано со схемой
                Log.e(TAG, "❌ Failed to read from disk cache: ${e.message}")
                cacheFile.delete()
            }
        }

        Log.d(TAG, "❌ Cache MISS: $key")
        return null
    }

    /**
     * Сохранить данные в кэш
     */
    inline fun <reified T> set(key: String, value: T) {
        try {
            val data = Json.encodeToString(value)
            val entry = CacheEntry(data, System.currentTimeMillis())

            // 1. Сохраняем в memory cache
            memoryCache[key] = entry

            // 2. Сохраняем на disk
            val cacheFile = diskCacheDir.resolve(key.hashCode().toString())
            val entryJson = Json.encodeToString(entry)
            cacheFile.writeText(entryJson)

            Log.d(TAG, "💾 Cached: $key (${data.length} bytes)")

            // 3. Проверяем размер disk cache и чистим если превышен лимит
            cleanupDiskCacheIfNeeded()
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to cache: $key - ${e.message}")
        }
    }

    /**
     * Очистить весь кэш
     */
    fun clearAll() {
        // Очищаем memory cache
        memoryCache.clear()

        // Очищаем disk cache
        diskCacheDir.listFiles()?.forEach { it.delete() }

        Log.i(TAG, "🗑️ Cache cleared")
    }

    /**
     * Удалить конкретную запись из кэша
     */
    fun remove(key: String) {
        // Удаляем из memory cache
        memoryCache.remove(key)

        // Удаляем с disk
        val cacheFile = diskCacheDir.resolve(key.hashCode().toString())
        cacheFile.delete()

        Log.d(TAG, "🗑️ Removed from cache: $key")
    }

    /**
     * Проверить и очистить disk cache если превышен лимит
     */
    @PublishedApi
    internal fun cleanupDiskCacheIfNeeded() {
        try {
            val files = diskCacheDir.listFiles() ?: return

            // Подсчитываем общий размер
            var totalSize = files.sumOf { it.length() }

            // Если превышен лимит - удаляем самые старые файлы
            if (totalSize > DISK_CACHE_LIMIT) {
                Log.w(TAG, "⚠️ Disk cache exceeded limit (${totalSize / 1024 / 1024}MB / ${DISK_CACHE_LIMIT / 1024 / 1024}MB)")

                // Сортируем по времени изменения (самые старые первые)
                val sortedFiles = files.sortedBy { it.lastModified() }

                var sizeToRemove = totalSize - DISK_CACHE_LIMIT
                var removedCount = 0

                for (file in sortedFiles) {
                    if (sizeToRemove <= 0) break

                    sizeToRemove -= file.length()
                    file.delete()
                    removedCount++
                }

                Log.i(TAG, "🗑️ Removed $removedCount old cache entries")
            }
        } catch (e: Exception) {
            Log.e(TAG, "❌ Failed to cleanup disk cache: ${e.message}")
        }
    }
}

/**
 * Запись кэша (данные + timestamp)
 */
@kotlinx.serialization.Serializable
@PublishedApi
internal data class CacheEntry(
    val data: String,
    val timestamp: Long
)
